internal/sdkio/ringbuffer_test.go

```diff
@@ -0,0 +1,213 @@
+			expectedWrittenBuffer: []byte("hello world"),
+		},
+		{
+			name:                  "No Bytes written",
+			capacity:              11,
+			input:                 []byte(""),
+			expectedStart:         0,
+			expectedEnd:           0,
+			expectedSize:          0,
+			expectedWrittenBuffer: []byte(""),
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			ringBuffer := New(c.capacity)
+			ringBuffer.Write(c.input)
+
+			if e, a := c.capacity, len(ringBuffer.slice); e != a {
+				t.Errorf("expect capacity of the ring buffer to be %v , got %v", e, a)
+			}
+
+			if e, a := c.expectedSize, ringBuffer.size; e != a {
+				t.Errorf("expect default size to be %v , got %v", e, a)
+			}
+
+			if e, a := c.expectedStart, ringBuffer.start; e != a {
+				t.Errorf("expect deafult start to point to %v , got %v", e, a)
+			}
+
+			if e, a := c.expectedEnd, ringBuffer.end; e != a {
+				t.Errorf("expect default end to point to %v , got %v", e, a)
+			}
+
+			if e, a := c.expectedWrittenBuffer, ringBuffer.slice; !bytes.Contains(a, e) {
+				t.Errorf("expect written bytes to be %v , got %v", e, a)
+			}
+		})
+	}
+
+}
+
+func TestRingBuffer_Read(t *testing.T) {
+	cases := []struct {
+		name                  string
+		capacity              int
+		input                 []byte
+		sizeofReadBytes       int
+		expectedStart         int
+		expectedEnd           int
+		expectedSize          int
+		expectedWrittenBuffer []byte
+		expectedReadByteSlice []byte
+	}{
+		{
+			name:                  "Capacity matches Bytes written",
+			capacity:              11,
+			input:                 []byte("Hello world"),
+			expectedStart:         0,
+			expectedEnd:           11,
+			expectedSize:          11,
+			sizeofReadBytes:       11,
+			expectedReadByteSlice: []byte("Hello world"),
+		},
+		{
+			name:                  "Capacity is lower than Bytes written",
+			capacity:              9,
+			input:                 []byte("hello world"),
+			expectedStart:         2,
+			expectedEnd:           2,
+			expectedSize:          9,
+			sizeofReadBytes:       9,
+			expectedReadByteSlice: []byte("llo world"),
+		},
+		{
+			name:                  "Capacity is more than Bytes written",
+			capacity:              15,
+			input:                 []byte("hello world"),
+			expectedStart:         0,
+			expectedEnd:           11,
+			expectedSize:          11,
+			sizeofReadBytes:       15,
+			expectedReadByteSlice: []byte("hello world"),
+		},
+		{
+			name:                  "No Bytes written",
+			capacity:              11,
+			input:                 []byte(""),
+			expectedStart:         0,
+			expectedEnd:           0,
+			expectedSize:          0,
+			sizeofReadBytes:       0,
+			expectedReadByteSlice: []byte(""),
+		},
+	}
+	for _, c := range cases {
+		t.Run(c.name, func(t *testing.T) {
+			ringBuffer := New(c.capacity)
+			readSlice := make([]byte, c.sizeofReadBytes)
+
+			ringBuffer.Write(c.input)
+			_, err := ringBuffer.Read(readSlice)
```

Comment:
We'll want to add a couple additional test cases here. 

* Read larger than buffer
* Read smaller than buffer
* Multiple reads, `e.g. io.Copy`
* Validation the io.EOF is returned only once the ring buffer has been fully read

Reply:
Probably also want a test case that does both write and read interwoven